src/background/sidebar-injector.ts

```diff
@@ -0,0 +1,485 @@
+     * this method.
+     *
+     * @param tab - A tab object representing the tab to insert the sidebar into.
+     * @param config - An object containing configuration info that is passed to
+     *   the app when it loads.
+     *
+     * Returns a promise that will be resolved if the injection succeeded
+     * otherwise it will be rejected with an error.
+     */
+    this.injectIntoTab = (tab: chrome.tabs.Tab, config: object = {}) => {
+      const tab_ = checkTab(tab);
+      if (isFileURL(tab_.url)) {
+        return injectIntoLocalDocument(tab_);
+      } else {
+        return injectIntoRemoteDocument(tab_, config);
+      }
+    };
+
+    /**
+     * Removes the Hypothesis sidebar from the tab provided.
+     *
+     * Returns a promise that will be resolved if the removal succeeded
+     * otherwise it will be rejected with an error.
+     */
+    this.removeFromTab = (tab: chrome.tabs.Tab) => {
+      const tab_ = checkTab(tab);
+      if (isPDFViewerURL(tab_.url)) {
+        return removeFromPDF(tab_);
+      } else if (isVitalSourceURL(tab_.url)) {
+        return removeFromVitalSource(tab_);
+      } else {
+        return removeFromHTML(tab_);
+      }
+    };
+
+    /**
+     * Request additional permissions that are required to inject Hypothesis
+     * into a given tab.
+     *
+     * Ideally the permissions request would just be part of {@link injectIntoTab}
+     * however it needs to be performed immediately after the user clicks the
+     * extension's toolbar icon, before any async calls, otherwise it will fail
+     * due to lack of a user gesture. See https://bugs.chromium.org/p/chromium/issues/detail?id=1363490.
+     */
+    this.requestExtraPermissionsForTab = async (tab: chrome.tabs.Tab) => {
+      const tab_ = checkTab(tab);
+      if (isVitalSourceURL(tab_.url)) {
+        return await chromeAPI.permissions.request({
+          permissions: ['webNavigation'],
+        });
+      } else {
+        // No extra permissions needed for other tabs.
+        return true;
+      }
+    };
+
+    function getPDFViewerURL(url: string) {
+      // Encode the original URL but preserve the fragment, so that a
+      // '#annotations' fragment in the original URL will persist and trigger the
+      // sidebar to focus and scroll to that annotation when the PDF viewer loads.
+      const parsedURL = new URL(url);
+      const hash = parsedURL.hash;
+      parsedURL.hash = '';
+      const encodedURL = encodeURIComponent(parsedURL.href);
+      return pdfViewerBaseURL + '?file=' + encodedURL + hash;
+    }
+
+    /**
+     * Returns true if the extension is permitted to inject a content script into
+     * a tab with a given URL.
+     */
+    async function canInjectScript(url: string) {
+      if (isSupportedURL(url)) {
+        return true;
+      } else if (isFileURL(url)) {
+        return chromeAPI.extension.isAllowedFileSchemeAccess();
+      } else {
+        return false;
+      }
```

Comment:
I'm not completely sure if this is equivalent, but I think this could be:

```suggestion
      if (isFileURL(url)) {
        return chromeAPI.extension.isAllowedFileSchemeAccess();
      }

      return isSupportedURL(url);
```

But this is not related with the TS migration itself.